doc/memos/rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md

```diff
@@ -0,0 +1,139 @@
+  storage devices
+
+## Status
+This document is currently under open discussion. This document is a product of
+[Configuration Task
+Force](https://github.com/RIOT-OS/RIOT/wiki/Configuration-Task-Force-(CTF)), and
+aims to describe the architecture of a runtime configuration system. The content
+of this document is licensed with a Creative Commons CC-BY-SA license.
+
+## Terminology
+This memo uses the [RFC2119](https://www.ietf.org/rfc/rfc2119.txt) terminology
+and the following acronyms:
+
+- RDM: RIOT Developer Memo
+- CTF: RIOT's Configuration Task Force
+- RCS: Runtime Configuration System
+
+# 1. Introduction
+This document specifies the proposed architecture by the
+Configuration Task Force (CFT) to implement a secure and reliable Runtime
+Configuration System (RCS), focusing in modularity, reusing of existing
+technologies in RIOT (network stack, storage interface) and adhering to
+standards.
+
+# 2. Architecture
+The RCS is formed by [the RIOT Registry](3-the-riot-registry), one or more [Registry Handlers](4-registry-handlers), one or more
+[storage facilities](5-storage-facilities) and a [configuration manager](6-configuration-manager) with one or more interfaces. The
+runtime configuration manager could in the future be replaced by a generic
+manager.  ![](./files/rdm-draft-alamos-lanzieri-runtime-configuration-architecture/architecture.png "Runtime Configuration Architecture")
+
+# 3. The RIOT Registry
+The RIOT Registry exposes an API for handling 
+**persistent module configurations**. The RIOT registry
+allows, among others, to:
+- Get or set configuration parameters via registry handlers
+- Commit a configuration group via the given registry handler
+- Load and set configuration parameters from a persistent storage device
+- Export configuration parameters (e.g copy to a buffer, print, etc)
+
+The RIOT registry consists of:
+- A list of Registry Handlers
+- A list of one or more Storage Facilities sources
+- One Storage destination
+
+# 4. Registry handlers
+A Registry Handler represents a configuration group in the RIOT Registry.
+Examples of configuration groups could be _Application_, _LoRaWAN keys_, etc.
+
+A Registry Handler is defined at least by a name and a series of handlers for interacting with the configuration parameters of the group. These handlers are:
+- `set`: Sets a value to a configuration parameter. This handler MUST take care
+of any logic to apply the value (e.g. data validation) or cache it until the commit handler is called
+- `get`: Gets the current value of a configuration parameter
+- `commit`: To be called when configuration parameters have been loaded from storage. It could be use for special needed logic on applying configuration parameters (e.g dependencies)
+- `export`: Calls an `export function` for each configuration parameter, with
+its name and value. Depending on the behavior of `export function`, this can be used for printing out all configurations, saving them in a persistent storage,
+etc.
```

Comment:
??
APIs of Registry and Registry Handlers seem overlapping??

Reply:
A developer should not call registry handlers by hand but use the orchestrator on top (Registry). Although they have a similar API, they are not intended to be used in the same way.

Think e.g about `netdev` and `netapi`. A developer would call `gnrc_netapi_send` to send a packet. `netdev->send` will be called as a result, but not intended to be called in higher layers.